### PR TITLE
Avoid race when scopeValidation is enabled

### PR DESCRIPTION
**Pull request by pakrym**
Fixes: https://github.com/aspnet/DependencyInjection/issues/467
@halter73 @mikeharder

### CODE REVIEW

File: src/Microsoft.Extensions.DependencyInjection/ServiceLookup/CallSiteValidator.cs [hunk 1]
```diff
@@ -2,21 +2,22 @@
...
-        private readonly Dictionary<Type, Type> _scopedServices = new Dictionary<Type, Type>();
+        private readonly ConcurrentDictionary<Type, Type> _scopedServices = new ConcurrentDictionary<Type, Type>();
 
         public void ValidateCallSite(Type serviceType, IServiceCallSite callSite)
         {
             var scoped = VisitCallSite(callSite, default(CallSiteValidatorState));
             if (scoped != null)
             {
-                _scopedServices.Add(serviceType, scoped);
+                _scopedServices[serviceType] = scoped;
```
Comment (mikeharder) [inline src/Microsoft.Extensions.DependencyInjection/ServiceLookup/CallSiteValidator.cs]: Is it correct to update the value if the key already exists?  Or would it be more correct to use `TryAdd()`, so the value is *not* updated if the key already exists?  Or does it not matter and either would be equally correct?

**Review by mikeharder — CHANGES_REQUESTED** [review]

File: src/Microsoft.Extensions.DependencyInjection/ServiceLookup/CallSiteValidator.cs (hunk 1 above)
Comment (pakrym) [inline src/Microsoft.Extensions.DependencyInjection/ServiceLookup/CallSiteValidator.cs]: We are displaying only one scoped service of multiple possible so I don't think there is a difference which one exactly.

**Review by mikeharder — APPROVED** [review]

**Comment by DamianEdwards** [comment]
I don't know if there's any runway left for 1.1.1. @Eilon?

**Comment by halter73** [comment]
Both a pro and con of merging this is that scopeValidation is opt-in. Presumably not many people are calling `collection.BuildServiceProvider(validateScopes: true)`.

**Comment by davidfowl** [comment]
The problem is that there's no workaround if you do want to use it. That kinda bug IMO classifies as patchable. 1.1.2 maybe?

**Comment by DamianEdwards** [comment]
Yeah I think it should be considered. I think it'll have to wait for 1.1.2 though

**Comment by Eilon** [comment]
Too late for the current patch, for sure. Can someone log a new bug for DI's next milestone so that we can consider? Please include some reasonable justification in terms of customer scenario.

**Pull request merged**